File: clients/java/src/org/jd3/D3FileTcp.java
 /* 
 * D3FileTcp.java - The implementation of a D3File who use a TCP/IP socket to communicate with the D3 server
 * Copyright (C) 2000 Christophe Marchal 
 * mccricri@jd3.sourceforges.com
 * http://jd3.sourceforge.com/ 
 * 
 * This program is free software; you can redistribute it and/or 
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation; either version 2 
 * of the License, or any later version. 
 * 
 * This program is distributed in the hope that it will be useful, 
 * but WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 * GNU General Public License for more details. 
 * 
 * You should have received a copy of the GNU General Public License 
 * along with this program; if not, write to the Free Software 
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA. 
 */

package org.jd3;

import java.io.*;
import java.util.*;

import org.jd3.utils.*;
import org.jd3.exceptions.*;

 /**
 * The implementation of a D3File who use a TCP/IP socket to communicate with the D3 server
 * @author Christophe Marchal
 * @version 0.8
 */
public class D3FileTcp implements D3File{

  /**
  * The connection to the D3 server
  */
  private D3Connection con;

  /**
  * the file descriptor
  */
  private int fd;

  /**
  * The name of the file
  */
  private String filename;
  
  /**
  * The name of the account where is the file
  */
  private String accountname;
  
  /**
  * Open a file store on the D3 server
  * @param pcon The conneciton to the server
  * @param paccount The account name where is store the file
  * @param pfilename The name of the file to open
  */
  public D3FileTcp(D3Connection pcon, String paccount, String pfilename) throws D3Exception{
    D3Buffer buf;
    this.con = pcon;
    fd = 0;
    filename = pfilename;
    accountname = paccount;
    
    buf = this.con.doit(D3Constants._D3_OPEN + "\001" + paccount + "\001" + pfilename + "\001");
    if(buf.getStatus() != D3Constants.D3_OK ) {
      throw new D3Exception("Erreur d'ouverture de " + paccount + "," + pfilename + ",");
    }

    try{
      fd = Integer.parseInt((String)buf.getData().elementAt(0));
    }
    catch(NumberFormatException e){
      throw new D3Exception("Erreur d'ouverture de " + paccount+ "," + pfilename + ",");
    }
  }

  /**
  * Open a file store on the D3 server using the account where the server is running
  * @param pcon The conneciton to the server
  * @param pfilename The name of the file to open
  */
  public D3FileTcp(D3Connection pcon, String pfilename) throws D3Exception{
       this(pcon, "", pfilename);
  }

  /**
  * Get the name of the file
  * @return The name of the file
  */
  public String getFileName() {
     return this.filename;
  }
  
  /**
  * Get the name of the accoutn where is the file
  * @return The name of the account where is the file
  */
  public String getAccountName() {
     return this.accountname;
  }
  
  /**
  * Lecture d'un record
  * @param key la clé
  * @param PItem le record à mettre à jour, doit être créé avant l'appel (éventuellement vide)
  * @return résultat de la lecture
  *     READ_FILENOTOPEN   Le fichier n'est pas ouvert, le record est mis à vide
  *     READ_RECORDEMPTY   le record n'existe pas pour la clé demandée, il est mis à vide
  *     READ_RECORDLOCKED  le record est déjà locké, dans ce cas le record contient
  *                          en <1> : le numéro de port qui lock
  *                          en <2> : le code user qui lock (si il est connu)
  *     D3_OK              si tout va bien, le record est mis à jour avec les infos
  */
  public synchronized int read(String key, D3Item PItem) throws D3Exception{
    D3Buffer buf = this.con.doit(D3Constants._D3_READ + "\001" + fd + "\001" + key + "\001");
    PItem.setRecord((String)buf.getData().elementAt(0));
    return buf.getStatus();
  }

  /**
  * Lecture d'un attribut d'un record
  * @param key  la clé du record à lire
  * @param am   la position de l'attribut à lire
  * @return  String la valeur de l'attribut lu, vide s'il y a eu un problème ou s'il est vide
  */
  public synchronized String readv(String key,int am) throws D3Exception{
    D3Buffer buf = this.con.doit(D3Constants._D3_READV + "\001" + fd + "\001" + key + "\001" + Integer.toString(am) + "\001");
    if(buf.getStatus() == D3Constants.D3_OK ) {
       return (String)buf.getData().elementAt(0);
    }else{
       return "";
    }
  }

  /**
  * Lecture et lock d'un record
  * @param key la clé
  * @param PItem le record à mettre à jour, doit être créé avant l'appel (éventuellement vide)
  * @return résultat de la lecture
  *     READ_FILENOTOPEN   Le fichier n'est pas ouvert, le record est mis à vide
  *     READ_RECORDEMPTY   le record n'existe pas pour la clé demandée, il est mis à vide
  *     READ_RECORDLOCKED  le record est déjà locké, dans ce cas le record contient
  *                          en <1> : le numéro de port qui lock
  *                          en <2> : le code user qui lock (si il est connu)
  *     D3_OK              si tout va bien, le record est mis à jour avec les infos
  */
  public synchronized int readu(String key, D3Item PItem) throws D3Exception{
    D3Buffer buf = this.con.doit(D3Constants._D3_READU + "\001" + fd + "\001" + key + "\001");
    if( buf.getStatus() == D3Constants.D3_OK){
       PItem.setRecord((String)buf.getData().elementAt(0));
    }
    return buf.getStatus();
  }

  /**
  * Ecrire un record
  * @param key la clé
  * @param PItem le record
  * @return int résultat de l'écriture
  */
  public synchronized int write(String key, D3Item PItem)  throws D3Exception{
    D3Buffer buf = this.con.doit(D3Constants._D3_WRITE + "\001" + fd + "\001" + key + "\001" + PItem.toString() + "\001");
    return buf.getStatus();
  }

  /**
  * Effacer un record
  * @param key la clé
  * @return int résultat du delete
  */
  public synchronized int delete(String key) throws D3Exception{
    D3Buffer buf = this.con.doit(D3Constants._D3_DELETE + "\001" + fd + "\001" + key + "\001");
    return buf.getStatus();
  }

  /**
  * Tester l'existance d'un record dans le fichier
  * @param key la clé
  * @return boolean false si le record n'existe pas ou s'il y a eu un problème
  *                 true  si le record existe
  */
  public synchronized boolean testExist(String key) throws D3Exception{
    D3Buffer buf = this.con.doit(D3Constants._D3_TEST_EXIST + "\001" + fd + "\001" + key + "\001");
    if ( buf.getStatus() == D3Constants.D3_OK){
      return ((String)buf.getData().elementAt(0)).equals("0");
    }else {
      return false;
    }
  }

  /**
  * Relacher un verrou
  * @param key la clé du verrou à lacher
  * @return int résultat de la commande
  */
  public synchronized int release(String key) throws D3Exception{
    D3Buffer buf = this.con.doit(D3Constants._D3_RELEASE + "\001" + fd + "\001" + key + "\001");
    return buf.getStatus();
  }

  /**
  * Effectue un select sur un fichier
  * @return int résultat de la commande
  */
  public synchronized D3SelectList select() throws D3Exception{
    int nb = 0;
    D3Buffer buf = this.con.doit(D3Constants._D3_SELECT + "\001" + fd + "\001");
    nb = Integer.parseInt((String)buf.getData().elementAt(1));
    return new D3SelectListFile(this.con,(String)buf.getData().elementAt(0),nb);
  }

}

File: clients/csharp/src/jd3/D3FileTcp.cs
/* 
* D3FileTcp.java - The implementation of a D3File who use a TCP/IP socket to communicate with the D3 server
* Copyright (C) 2000 Christophe Marchal 
* mccricri@jd3.sourceforges.com
* http://jd3.sourceforge.com/ 
* 
* This program is free software; you can redistribute it and/or 
* modify it under the terms of the GNU General Public License 
* as published by the Free Software Foundation; either version 2 
* of the License, or any later version. 
* 
* This program is distributed in the hope that it will be useful, 
* but WITHOUT ANY WARRANTY; without even the implied warranty of 
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
* GNU General Public License for more details. 
* 
* You should have received a copy of the GNU General Public License 
* along with this program; if not, write to the Free Software 
* Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA. 
*/
namespace org.jd3
{
	using System;
	using org.jd3.utils;
	using org.jd3.exceptions;
	
	/// <summary> The implementation of a D3File who use a TCP/IP socket to communicate with the D3 server
	/// </summary>
	/// <author> Christophe Marchal
	/// </author>
	/// <version> 0.8
	/// 
	/// </version>
	public class D3FileTcp : D3File
	{
		public virtual System.String FileName
		{
			get
			{
				return this.filename;
			}
			
		}
		public virtual System.String AccountName
		{
			get
			{
				return this.accountname;
			}
			
		}
		
		/// <summary> The connection to the D3 server
		/// </summary>
		private D3Connection con;
		
		/// <summary> the file descriptor
		/// </summary>
		private int fd;
		
		/// <summary> The name of the file
		/// </summary>
		private System.String filename;
		
		/// <summary> The name of the account where is the file
		/// </summary>
		private System.String accountname;
		
		/// <summary> Open a file store on the D3 server
		/// </summary>
		/// <param name="pcon">The conneciton to the server
		/// </param>
		/// <param name="paccount">The account name where is store the file
		/// </param>
		/// <param name="pfilename">The name of the file to open
		/// 
		/// </param>
		public D3FileTcp(D3Connection pcon, System.String paccount, System.String pfilename)
		{
			D3Buffer buf;
			this.con = pcon;
			fd = 0;
			filename = pfilename;
			accountname = paccount;
			
			buf = this.con.doit(D3Constants._D3_OPEN + "\x0001" + paccount + "\x0001" + pfilename + "\x0001");
			if (buf.Status != D3Constants.D3_OK)
			{
				throw new D3Exception("Erreur d'ouverture de " + paccount + "," + pfilename + ",");
			}
			
			try
			{
				fd = System.Int32.Parse((System.String) buf.Data[0]);
			}
			catch (System.FormatException e)
			{
				throw new D3Exception("Erreur d'ouverture de " + paccount + "," + pfilename + "," + e.ToString());
			}
		}
		
		/// <summary> Open a file store on the D3 server using the account where the server is running
		/// </summary>
		/// <param name="pcon">The conneciton to the server
		/// </param>
		/// <param name="pfilename">The name of the file to open
		/// 
		/// </param>
		public D3FileTcp(D3Connection pcon, System.String pfilename):this(pcon, "", pfilename)
		{
		}
		
		/// <summary> Get the name of the file
		/// </summary>
		/// <returns>The name of the file
		/// 
		/// </returns>
		
		/// <summary> Get the name of the accoutn where is the file
		/// </summary>
		/// <returns>The name of the account where is the file
		/// 
		/// </returns>
		
		/// <summary> Lecture d'un record
		/// </summary>
		/// <param name="key">la cl
		/// </param>
		/// <param name="PItem">le record  mettre  jour, doit tre cr avant l'appel (ventuellement vide)
		/// </param>
		/// <returns>rsultat de la lecture
		/// READ_FILENOTOPEN   Le fichier n'est pas ouvert, le record est mis  vide
		/// READ_RECORDEMPTY   le record n'existe pas pour la cl demande, il est mis  vide
		/// READ_RECORDLOCKED  le record est dj lock, dans ce cas le record contient
		/// en <1> : le numro de port qui lock
		/// en <2> : le code user qui lock (si il est connu)
		/// D3_OK              si tout va bien, le record est mis  jour avec les infos
		/// 
		/// </returns>
		//UPGRADE_NOTE: Synchronized keyword was removed from method 'read'. Lock expression was added. 'ms-help://MS.VSCC/commoner/redir/redirect.htm?keyword="jlca1027"'
		public virtual int read(System.String key, D3Item PItem)
		{
			lock(this)
			{
				D3Buffer buf = this.con.doit(D3Constants._D3_READ + "\x0001" + fd + "\x0001" + key + "\x0001");
				PItem.Record = (System.String) buf.Data[0];
				return buf.Status;
			}
		}
		
		/// <summary> Lecture d'un attribut d'un record
		/// </summary>
		/// <param name="key"> la cl du record  lire
		/// </param>
		/// <param name="am">  la position de l'attribut  lire
		/// </param>
		/// <returns> String la valeur de l'attribut lu, vide s'il y a eu un problme ou s'il est vide
		/// 
		/// </returns>
		//UPGRADE_NOTE: Synchronized keyword was removed from method 'readv'. Lock expression was added. 'ms-help://MS.VSCC/commoner/redir/redirect.htm?keyword="jlca1027"'
		public virtual System.String readv(System.String key, int am)
		{
			lock(this)
			{
				D3Buffer buf = this.con.doit(D3Constants._D3_READV + "\x0001" + fd + "\x0001" + key + "\x0001" + System.Convert.ToString(am) + "\x0001");
				if (buf.Status == D3Constants.D3_OK)
				{
					return (System.String) buf.Data[0];
				}
				else
				{
					return "";
				}
			}
		}
		
		/// <summary> Lecture et lock d'un record
		/// </summary>
		/// <param name="key">la cl
		/// </param>
		/// <param name="PItem">le record  mettre  jour, doit tre cr avant l'appel (ventuellement vide)
		/// </param>
		/// <returns>rsultat de la lecture
		/// READ_FILENOTOPEN   Le fichier n'est pas ouvert, le record est mis  vide
		/// READ_RECORDEMPTY   le record n'existe pas pour la cl demande, il est mis  vide
		/// READ_RECORDLOCKED  le record est dj lock, dans ce cas le record contient
		/// en <1> : le numro de port qui lock
		/// en <2> : le code user qui lock (si il est connu)
		/// D3_OK              si tout va bien, le record est mis  jour avec les infos
		/// 
		/// </returns>
		//UPGRADE_NOTE: Synchronized keyword was removed from method 'readu'. Lock expression was added. 'ms-help://MS.VSCC/commoner/redir/redirect.htm?keyword="jlca1027"'
		public virtual int readu(System.String key, D3Item PItem)
		{
			lock(this)
			{
				D3Buffer buf = this.con.doit(D3Constants._D3_READU + "\x0001" + fd + "\x0001" + key + "\x0001");
				if (buf.Status == D3Constants.D3_OK)
				{
					PItem.Record = (System.String) buf.Data[0];
				}
				return buf.Status;
			}
		}
		
		/// <summary> Ecrire un record
		/// </summary>
		/// <param name="key">la cl
		/// </param>
		/// <param name="PItem">le record
		/// </param>
		/// <returns>int rsultat de l'criture
		/// 
		/// </returns>
		//UPGRADE_NOTE: Synchronized keyword was removed from method 'write'. Lock expression was added. 'ms-help://MS.VSCC/commoner/redir/redirect.htm?keyword="jlca1027"'
		public virtual int write(System.String key, D3Item PItem)
		{
			lock(this)
			{
				//UPGRADE_TODO: The equivalent in .NET for method 'java.Object.toString' may return a different value. 'ms-help://MS.VSCC/commoner/redir/redirect.htm?keyword="jlca1043"'
				D3Buffer buf = this.con.doit(D3Constants._D3_WRITE + "\x0001" + fd + "\x0001" + key + "\x0001" + PItem.ToString() + "\x0001");
				return buf.Status;
			}
		}
		
		/// <summary> Effacer un record
		/// </summary>
		/// <param name="key">la cl
		/// </param>
		/// <returns>int rsultat du delete
		/// 
		/// </returns>
		//UPGRADE_NOTE: Synchronized keyword was removed from method 'delete'. Lock expression was added. 'ms-help://MS.VSCC/commoner/redir/redirect.htm?keyword="jlca1027"'
		public virtual int delete(System.String key)
		{
			lock(this)
			{
				D3Buffer buf = this.con.doit(D3Constants._D3_DELETE + "\x0001" + fd + "\x0001" + key + "\x0001");
				return buf.Status;
			}
		}
		
		/// <summary> Tester l'existance d'un record dans le fichier
		/// </summary>
		/// <param name="key">la cl
		/// </param>
		/// <returns>boolean false si le record n'existe pas ou s'il y a eu un problme
		/// true  si le record existe
		/// 
		/// </returns>
		//UPGRADE_NOTE: Synchronized keyword was removed from method 'testExist'. Lock expression was added. 'ms-help://MS.VSCC/commoner/redir/redirect.htm?keyword="jlca1027"'
		public virtual bool testExist(System.String key)
		{
			lock(this)
			{
				D3Buffer buf = this.con.doit(D3Constants._D3_TEST_EXIST + "\x0001" + fd + "\x0001" + key + "\x0001");
				if (buf.Status == D3Constants.D3_OK)
				{
					return ((System.String) buf.Data[0]).Equals("0");
				}
				else
				{
					return false;
				}
			}
		}
		
		/// <summary> Relacher un verrou
		/// </summary>
		/// <param name="key">la cl du verrou  lacher
		/// </param>
		/// <returns>int rsultat de la commande
		/// 
		/// </returns>
		//UPGRADE_NOTE: Synchronized keyword was removed from method 'release'. Lock expression was added. 'ms-help://MS.VSCC/commoner/redir/redirect.htm?keyword="jlca1027"'
		public virtual int release(System.String key)
		{
			lock(this)
			{
				D3Buffer buf = this.con.doit(D3Constants._D3_RELEASE + "\x0001" + fd + "\x0001" + key + "\x0001");
				return buf.Status;
			}
		}
		
		/// <summary> Effectue un select sur un fichier
		/// </summary>
		/// <returns>int rsultat de la commande
		/// 
		/// </returns>
		//UPGRADE_NOTE: Synchronized keyword was removed from method 'select'. Lock expression was added. 'ms-help://MS.VSCC/commoner/redir/redirect.htm?keyword="jlca1027"'
		public virtual D3SelectList select()
		{
			lock(this)
			{
				int nb = 0;
				D3Buffer buf = this.con.doit(D3Constants._D3_SELECT + "\x0001" + fd + "\x0001");
				nb = System.Int32.Parse((System.String) buf.Data[1]);
				return new D3SelectListFile(this.con, (System.String) buf.Data[0], nb);
			}
		}
	}
}
